### ISSUE TITLE
tsOffset may fail depending on the Intl.DateTimeFormat implementation

### ISSUE THREAD

**Issue by martin-trummer** [date-fns/tz#23]
https://github.com/date-fns/tz/blob/213903702d7c5fcd4f01479ba7370fe917195a65/src/tzOffset/index.ts#L18-L23

The problem is that the [format](url) function is not bound to the class. So if the implementation for [format()](url) uses [this](url), it will fail at runtime.

We can see the issue, when we use the [Time Travel Chrom Extension](https://chromewebstore.google.com/detail/time-travel/jfdbpgcmmenmelcghpbbkldkcfiejcjg) which is very useful for testing

possible fix would be
```typescript
function getBoundFormat(timeZone: string) {
  const dateTimeFormat = new Intl.DateTimeFormat("en-GB", {
    timeZone,
    hour: "numeric",
    timeZoneName: "longOffset"
  })
  return dateTimeFormat.format.bind(dateTimeFormat);
}

export function tzOffset(timeZone, date) {
  try {
    const format = offsetFormatCache[timeZone] ||= getBoundFormat(timeZone);
```

**Comment by Perdolique** [comment]
@kossnocorp which browsers are supported? I see that `longOffset` isn't supported in Safari below 15.4 and Chrome below 95, for example, but can't find a compatibility table in the documentation.

**Comment by ryaa** [comment]
@Perdolique 

> I see that `longOffset` isn't supported in Safari below 15.4 and Chrome below 95, for example, but can't find a compatibility table in the documentation.

Yes, it seems to be a limitation in `date-fns/tz` package. The problem is that in `tzOffset` method the `Intl.DateTimeFormat` is constructed with `options.timeZoneName` set to **longOffset** (https://github.com/date-fns/tz/blob/main/src/tzOffset/index.ts#L22) which is supported in Safari 15.4+ (iOS 15.4 and above) - see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat#browser_compatibility or below Image

<img width="830" alt="Image" src="https://github.com/user-attachments/assets/092089d3-5de0-4faa-b9db-f0f40e000c60" />

**Comment by ryaa** [comment]
@csandman 
> I'm currently looking into [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) for a polyfill approach, but I'm not sure the best way to actually test on one of these older browsers.

Have you managed to resolve this problem with older browsers with [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) polyfills?
The alternative could be to make the change in https://github.com/date-fns/tz/blob/main/src/tzOffset/index.ts#L22 to feature-detect support for `longOffset` and use `long` in Safari 15.3 and below and `longOffset` in 15.4+

**Comment by ryaa** [comment]
I can confirm that [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) polyfills does not fix the problem when formatting the date on iOS 15.4 or below. The error is `RangeError: timeZoneName must be "short" or "long" and happens because `timeZoneName: "longOffset"` which is not supported on iOS 15.4 or below - see https://github.com/date-fns/tz/blob/main/src/tzOffset/index.ts#L22

Please also see https://github.com/date-fns/tz/issues/38 and https://github.com/formatjs/formatjs/issues/4804

**Comment by ryaa** [comment]
Until [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) polyfills is fixed to correctly handle the above use case, I implemented a temporary fix for `date-fns/tz` package. The fix can be applied using [patch-package](https://github.com/ds300/patch-package) and the attached file (copy the file to the appropriate director and run `npm install` to apply the patch)

[@date-fns+tz+1.2.0.patch](https://github.com/user-attachments/files/18845298/%40date-fns%2Btz%2B1.2.0.patch)

**Comment by csandman** [comment]
@ryaa Glad you found a workaround, sorry I meant to reply saying I didn't have any luck getting formatjs to do what I wanted either haha. I'll take a look into your patch! Do you have a snippet of the exact code you changed?

**Comment by ryaa** [comment]
> [@ryaa](https://github.com/ryaa) I'll take a look into your patch! Do you have a snippet of the exact code you changed?

This patch will update two files `node_modules/@date-fns/tz/tzOffset/index.cjs` and `node_modules/@date-fns/tz/tzOffset/index.js`. The resulting files are attached, so you should be able to easily compare to see the difference (just unzip them)

[index.cjs.zip](https://github.com/user-attachments/files/18882530/index.cjs.zip)
[index.js.zip](https://github.com/user-attachments/files/18882531/index.js.zip)

> Also, does your patch actually allow dates to still be formatted in the specified timezone?

I believe yes. At least this is the requirement in my app and my patch seems to format the dates in any arbitrary timezone

**Comment by csandman** [comment]
> I believe yes. At least this is the requirement in my app and my patch seems to format the dates in any arbitrary timezone

Nice, is there any reason not to PR this against the main project then? It seems like it would be generally helpful for people.

**Comment by ryaa** [comment]
> Nice, is there any reason not to PR this against the main project then? It seems like it would be generally helpful for people.

The issue is with [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) polyfills and the fix should be provided there, not in `date-fns/tz`

**Comment by csandman** [comment]
> The issue is with [formatjs](https://formatjs.github.io/docs/polyfills/intl-datetimeformat#usage) polyfills and the fix should be provided there, not in `date-fns/tz`

I guess that depends on how you look at it, if this package could offer a fallback without requiring a polyfill, that could be a better option for some people.

**Comment by seravifer** [comment]
We have the same problem on Chrome 91, "longOffset" not support until Chrome 95. I know is an old version but our customer can't update because is custom Android version with fixed webview.

**Comment by parksb** [comment]
I think [ryaa’s pull request](https://github.com/date-fns/tz/pull/45) could resolve this issue. Are there any updates on it? It would be great if the member of date-fns could review it.

**Comment by samuelcole** [comment]
We are getting this exception quite frequently with our users on older browsers, would love to see the PR merged.

**Comment by rich-young-slalom** [comment]
Also found a related issue around longOffset and android on react-native. 

the longOffset format is returning "GMT+08:00 (hour: 14)" and then the naive split on ':' is creating an array with ["+08", "00 (hour", "14)"] and then the minute number conversion is returning NaN.

There is the existing regexp that is being used in the fallback if there's an error, but should probably be used always instead of the string split.

**Comment by kossnocorp** [comment]
The fix turned out to be much more complicated to pull off as it was really challenging to actually test it. I ended up using Hermes JS Engine and JavaScriptCore directly to reproduce, write, and test the fix, and it seems to work now. I couldn't run Vitest tests in it due to lack of OS interop, so I tested it only with a handful of tests rather than a full suite.

Please make sure to include the following [Format.JS polyfills](https://formatjs.github.io/docs/polyfills/intl-datetimeformat/) in the main entry point:

```ts
import "@formatjs/intl-getcanonicallocales/polyfill";
import "@formatjs/intl-locale/polyfill";
import "@formatjs/intl-pluralrules/polyfill";
import "@formatjs/intl-numberformat/polyfill";
import "@formatjs/intl-numberformat/locale-data/en";
import "@formatjs/intl-datetimeformat/polyfill";
import "@formatjs/intl-datetimeformat/locale-data/en";
import "@formatjs/intl-datetimeformat/add-golden-tz"; // or: "@formatjs/intl-datetimeformat/add-all-tz"
```

I published the fix it as [`@date-fns/tz@1.3.0`](https://github.com/date-fns/tz/releases/tag/v1.3.0).

Any issues related to exotic environments now can be addressed much easier thanks to [a script that tests engines directly](https://github.com/date-fns/tz/blob/main/scripts/test/engines.sh).

If you still have issues, please open a new issue with your system time zone, browser, or engine version (e.g., Hermes v0.12.0), the date time string you are working with and the time zone that fails.

**Comment by csandman** [comment]
@kossnocorp does your fix handle the case of older browsers specifically not provided `longOffset`? Or would the polyfills only load in if the date time format API is not available in general? Just curious because I know a lot of polyfills won't even be loaded if the API exists at all.

**Comment by csandman** [comment]
> @csandman not yet and I'm unsure if polyfills would help, but I will test it.
> 
> I'm working on a solution for unsupported `longOffset`, but it will require including a subset of tzdata as any other `timeZoneName` does not output offset consistently.
> 
> I created an issue to track the progress: https://github.com/date-fns/tz/issues/65

Cool, thanks for the update!
